### PR TITLE
[NEW] Add message action to copy message to input as reply

### PR DESCRIPTION
**Pull request by mrsimpson**
## Motivation

As a developer of a bot, I want to suggest a response to the user. Instead of just making the user to send a fully predefined message, the user shall adjust the message (think of it as a template for the reply).
This reduces typing overhead, reduces errors during the conversation and makes sure the user follows some expected structure

## Implementation

There already was a `message-action` to send a message by clicking a button living in an attachment.
This enhancement allows to create an action button (either from an integration, an API or a bot) which copies a predefined message as template into the message input (browser only!)

## Consumption

The button is part of an attachment's actions and can be created anywhere there is a message being created. 
Sample for an outgoing integration

```javascript
class Script{
  
    prepare_outgoing_request({ request }) {

      return {
        message: {
          text: "Would you be so kind to let me know what you did today?",
          attachments: [{
            actions: [
              {
                type: "button",
                msg_in_chat_window: true,
                msg_processing_type: "respondWithMessage",
                text: "Respond easily",
                msg: "**What I worked on today**\n- \n\n**What I am planning to do tomorrow**\n- "
              }
            ]
            }]
        }
      }
    }
}

```

## How it looks like

Sample: A Daily-Bot is gathering input from users:

## Open issues

Mobile adaption - @rafaelks since this is a client side action, the mobile apps would have to implement this copy-to-input as well.

### CODE REVIEW

**Comment by rafaelks** [comment]
@mrsimpson That looks pretty nice! Just curious, did you follow some pattern from other kind of bots in order to implement this? We definitely want this on the mobile! Any way to customize the button colors or something?

**Comment by mrsimpson** [comment]
@rafaelks the general "architecture" of the message actions is not invented by myself. 
The package that brought the ability to provide actionable buttons in attachments was #11473 .
To be honest, I don't like it much, we had something on our fork which I preferred, but this PR made it into the core.
I didn't find any API of e. g. a bot framework, which matches the structure, only documentation in a [wiki of a fork](https://github.com/WideChat/Rocket.Chat.Android/wiki/Phase-1--Rich-Messaging-Acceptance-Test) . @ear-dev or @bizzbyster seem to be driving this and may be able to answer. They obviously also have some rendering of the actions in their Android-App-Fork.

This PR only adds the ability to provide a reply-with-message-capability - and tries to be as compatible to the existing stuff merged with #11473 as possible.

**Comment by rafaelks** [comment]
@mrsimpson Got it, that makes perfect sense, thanks for explaining that. I can see @bizzbyster have a good documentation on their Wiki about this new properties and how they work, what do you guys thinking about adding the parts that Rocket.Chat core supports in our own docs already, so we keep it updated and will be easier to other clients to support the same functionalities? I know that our Android app already supports some of the actions, was a contribution from @bizzbyster team.

**Comment by wreiske** [comment]
@bizzbyster made a comment about better UI support without restoring to things like asterisks. Maybe a look at https://api.slack.com/dialogs will give some ideas?
image

**Comment by mrsimpson** [comment]
@wreiske I am quite sure I understood @bizzbyster 's comment about the better UI-support - and I agree: There should be better support for more controls like input- or comboboxes. There also are other features which need to be provided in order to make message-based rich interactions possible. 

But this is not the intention of this PR. The feature is a bit different: The purpose is to provide a button which creates a template for a message. Of course, this could be seen as an alternative implementation of a form, but it's so limited it's in fact a different feature. The sample I chose with the asterisks was only to illustrate there's an option to supply markdown in the template.

So: Yes, there should be other issues / PRs to provide real rich messaging.

**Comment by mrsimpson** [comment]
@timkinnane I couldn't agree more!

I believe the action that is take when clicking on a button is modeled very ... simply at the moment. This also makes it very hard to extend. 

My intention was to be as compatible as possible since I don't know which parts of the design were taken (by @bizzbyster 's team) from another spec with which the actions should be compatible and which one is *"invented here"*.

Therefore, I also [added default behavior](https://github.com/RocketChat/Rocket.Chat/pull/12626/files#diff-905538b5beeef04db294daf685105401R11) in order to be compatible with the current spec. Given the existing design, I didn't see another non-disruptive way (so that potential consumers of the API are not breaking).

**IF** we are free to change here, I'd do quite some changes. Before I go into the details, let me elaborate the model:

I understood that there are basically two levels of types:
1) Where's the action processed 
```typescript
enum processingSide {
    client,
    server,
}
```
2) Which is the directive to process it
```typescript
enum clientSideProcessingDirective {
   postMessage, // creates a fixed message issued by the executer
   messageInoutWithTemplate, // fills in the message input with a given template,
   httpRequest, // triggers a http-request from within the client (browser, native app)
   openForm, // opens a dialog with a rich UI
}

enum serverSideProcessingDirective {
   callMethod, // triggers the execution of a defined method in the context of the executor
   httpRequest, // triggers a server side http-request
}
```
In order to keep this flexible, I'd propose a command-pattern, so that the consumer creates an object describing the action, e. g. 
```javascript
{
  processingSide: 'client',
  command: {
    directive: 'postMessage',
    msg: 'booo!'
  }
}
```

Then, handlers for each command could be provided on each platform, effectively executing the directive. This was extensible, led to maintainable code and could accommodate all the existing requirements.

However, this was extremely incompatible with the existing codebase.
@bizzbyster @rafaelks @timkinnane thoughts?

**Comment by mrsimpson** [comment]
@RocketChat/core Any ideas / concepts about how to properly provide rich message attachments? I'm more than willing to contribute also a more powerful solution - but only once I can be sure, that this is going to be supported by the core (also on mobile)

**Comment by ggazzo** [comment]
@rodrigok could you check the tests?

**Comment by sampaiodiego** [comment]
> @RocketChat/core Any ideas / concepts about how to properly provide rich message attachments? I'm more than willing to contribute also a more powerful solution - but only once I can be sure, that this is going to be supported by the core (also on mobile)

yes! I have recently discussed with @engelgabriel about this subject.. he thinks we should have an API like [slack's blocks](https://api.slack.com/reference/messaging/blocks) .. so we'll definitely give rich messages a look soon.

in the meanwhile I think we could have this feature merged in as we'll have a **new schema** for rich messages in the feature.

File: packages/rocketchat-ui/client/views/app/room.js [hunk 1]
```diff
@@ -856,6 +856,18 @@ Template.room.events({
 			Meteor.call('sendMessage', msgObject);
 		});
 	},
+	'click .js-actionButton-respondWithMessage'(event) {
+		const msg = event.currentTarget.value;
+		if (!msg) {
+			return;
+		}
+
+		const input = $('.messages-container .rc-message-box .rc-message-box__textarea').get(0);
```
Comment (sampaiodiego) [inline packages/rocketchat-ui/client/views/app/room.js]: this should use `chatMessages` object, like [here](https://github.com/RocketChat/Rocket.Chat/blob/develop/app/ui-message/client/startup/restoreText.js#L18)

Comment (sampaiodiego) [inline packages/rocketchat-ui/client/views/app/room.js]: this should use `chatMessages` object, like [here](https://github.com/RocketChat/Rocket.Chat/blob/develop/app/ui-message/client/startup/restoreText.js#L18)

**Comment by mrsimpson** [comment]
@d-gubert i believe I introduced the timeout since the input was not always available to the jQuery-selector I hacked previously. The timeout was only there to be executed async. 
However, this may be solved by the new mechanism as suggested by Diego. 
Just wanted to explain

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by d-gubert** [comment]
Cool :)

We've tested it with some different input values and it worked without the timeout, so we've decided to remove it.



> However, this may be solved by the new mechanism as suggested by Diego.

Indeed this might have been the case :)

**Pull request merged**